File: articles/linear-algebra-for-systems-modeling/articles/case-study-network-system-modeling/calculators/python/network_system_calculator.py
from __future__ import annotations

import csv
import json
from pathlib import Path


def main() -> None:
    output_dir = Path("outputs")
    output_dir.mkdir(parents=True, exist_ok=True)

    result = {
        "calculator": "case_study_network_system_modeling_calculator",
        "workflow_name": "network_system_modeling_audit",
        "network_name": "synthetic_infrastructure_service_network",
        "node_count": 5,
        "edge_count": 6,
        "total_weight": 17.0,
        "highest_weighted_degree_node": "B",
        "highest_weighted_degree": 12.0,
        "laplacian_trace": 34.0,
        "baseline_component_count": 1,
        "stressed_component_count": 1,
        "removed_edge": "B-D",
        "warning": "Network metrics depend on node definitions, edge meanings, weights, directionality, boundary choices, and missing-edge assumptions."
    }

    with (output_dir / "case_study_network_system_modeling_calculator.json").open("w", encoding="utf-8") as handle:
        json.dump(result, handle, indent=2, sort_keys=True)

    with (output_dir / "case_study_network_system_modeling_calculator.csv").open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(result.keys()))
        writer.writeheader()
        writer.writerow(result)

    print(json.dumps(result, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: articles/linear-algebra-for-systems-modeling/articles/case-study-network-system-modeling/calculators/r/network_system_calculator.R
result <- data.frame(
  calculator = "case_study_network_system_modeling_calculator",
  workflow_name = "network_system_modeling_audit",
  network_name = "synthetic_infrastructure_service_network",
  node_count = 5,
  edge_count = 6,
  total_weight = 17.0,
  highest_weighted_degree_node = "B",
  highest_weighted_degree = 12.0,
  laplacian_trace = 34.0,
  baseline_component_count = 1,
  stressed_component_count = 1,
  removed_edge = "B-D",
  warning = "Network metrics depend on node definitions, edge meanings, weights, directionality, boundary choices, and missing-edge assumptions."
)

dir.create("outputs", recursive = TRUE, showWarnings = FALSE)
write.csv(result, "outputs/r_case_study_network_system_modeling_calculator.csv", row.names = FALSE)
print(result)
